### PR TITLE
feat(add ajv-errors support): add foundation

### LINKED ISSUE #1235: [FEAT] Add ajv-errors support
## Information

Add ajv-errors message customization support.

Ref: https://www.npmjs.com/package/ajv-errors
Related: #833

## Example

Configuration:
```js
const Ajv = require("ajv").default
const ajv = new Ajv({allErrors: true})
// Ajv option allErrors is required
require("ajv-errors")(ajv /*, {singleError: true} */)

const schema = {
  type: "object",
  required: ["foo"],
  properties: {
    foo: {type: "integer"},
  },
  additionalProperties: false,
  errorMessage: "should be an object with an integer property foo only",
}

const validate = ajv.compile(schema)
console.log(validate({foo: "a", bar: 2})) // false
console.log(validate.errors) // processed errors
```

Output:

```jsonc
[
  {
    keyword: "errorMessage",
    message: "should be an object with an integer property foo only",
    // ...
    params: {
      errors: [
        {keyword: "additionalProperties", dataPath: "" /* , ... */},
        {keyword: "type", dataPath: ".foo" /* , ... */},
      ],
    },
  },
]
```

### Example 1:
Expected json-schema:

```typescript
{
  type: "object",
  required: ["foo"],
  properties: {
    foo: {type: "integer"},
  },
  additionalProperties: false,
  errorMessage: {
    type: "should be an object", // will not replace internal "type" error for the property "foo"
    required: "should have property foo",
    additionalProperties: "should not have properties other than foo",
  },
}
```

Solution using decorator:

```typescript
import {ErrorMsg} from "@tsed/ajv";

@ErrorMsg({
   type: "should be an object", 
   required: "should have property foo",
   additionalProperties: "should not have properties other than foo"
})
export class MyModel {
    @Required()
    @Integer()
    foo: number;
}
```

## Example 2

For keywords "required" and "dependencies" it is possible to specify different messages for different properties.

Expected json-schema:

```js
{
  type: "object",
  required: ["foo", "bar"],
  properties: {
    foo: {type: "integer"},
    bar: {type: "string"},
  },
  errorMessage: {
    type: "should be an object", // will not replace internal "type" error for the property "foo"
    required: {
      foo: 'should have an integer property "foo"',
      bar: 'should have a string property "bar"',
    },
  },
}
```

Solution using decorator:

```typescript
import {ErrorMsg, TypeMsg, RequiredMsg} from "@tsed/ajv";

@TypeMsg("should be an object") // shortcut to ErrorMessage({type: 'msg'})
export class MyModel {
    @Required()
    @Integer()
    @RequiredError('integer', 'should have an integer property "foo"')
    foo: number;
   
   @Required()
   @RequiredError("required", 'should have a string property "bar"')
   bar: string;
}
```
alternative that need a huge rework, but which is more elegant:

```typescript
import {ErrorMsg, TypeMsg, RequiredMsg} from "@tsed/ajv";

@TypeError("should be an object") // shortcut to ErrorMessage({type: 'msg'})
export class MyModel {
    @Required()
    @Integer().Error( 'should have an integer property "foo"')

    foo: number;
   
   @Required().Error('is required')
   bar: string;
}
```

### Example 3 - Default message

When the value of keyword errorMessage is an object you can specify a message that will be used if any error appears that is not specified by keywords/properties/items using `_` property:

```js
const schema = {
  type: "object",
  required: ["foo", "bar"],
  properties: {
      foo: {type: "integer", minimum: 2},
      bar: {type: "string", minLength: 2},
  },
  additionalProperties: false,
  errorMessage: {
    properties: {
      foo: "data.foo should be integer >= 2",
      bar: "data.bar should be string with length >= 2",
    },
    _: 'data should have properties "foo" and "bar" only',
  },
}
```

Solution using decorator:

```typescript
import {TypeError, RequiredError, DefaultError} from "@tsed/ajv";

@DefaultMsg('data should have properties "foo" and "bar" only') // eq: ErrorMsg({_: "message"})
export class MyModel {
    @Required()
    @Integer()
    @RequiredError('should have an integer property "foo"')
    foo: number;
   
   @Required()
   @RequiredError('should have a string property "bar"')
   bar: string;
}
```

## Acceptance criteria

- [ ] Decorators works on class and properties
- [ ] Decorators are correctly exported:
   - [ ] ErrorMsg
   - [ ] TypeError
   - [ ] RequiredError
   - [ ] DefaultError
- [ ] Ajv is correctly configured with ajv-errors (Edit https://github.com/TypedProject/tsed/blob/production/packages/ajv/src/services/Ajv.ts#L48)
- [ ] Documentation is up-to-date (model.md and ajv.md)
- [ ] Unit test cover correctly the decorators.
- [ ] Integration test cover the success and error cases.

### PR DESCRIPTION
**Pull request by Crunchyman-ralph**
## Information

| Type                  | Breaking change |
| --------------------- | --------------- |
| Feature | No          |

---

- resolves #1235 
#### Customise error messages for decorators:
```typescript
import {Required} from "@tsed/schema";

class Model {
      @Required().Error("custom message")
      num: number;
}

```

#### Add a default ajv message:
```typescript
@DefaultMsg("foo should be a string")
    class Model {
      property: number;
    }
```

#### Add a more global ErrorMsg decorator (not recommended)
```typescript
class Model {
      @ErrorMsg({type: "foo should be a string"})
      property: number;
    }
```

#### Add a TypeError message
```typescript
@TypeError("foo should be a string")
    class Model {
      property: number;
    }
```
## Todos

- [x] Tests
- [x] e2e Tests
- [x] Coverage
- [x] Example
- [x] Documentation

### CODE REVIEW

**Review by Romakita — APPROVED** [review]

**Comment by Romakita** [comment]
@Crunchyman-ralph  Perfect ;)

**Comment by Crunchyman-ralph** [comment]
> * [x]  Decorators works on class and properties
> * [x]  Decorators are correctly exported:
>   
>   * [x]  ErrorMsg
>   * [x]  TypeError
>   * [x]  RequiredError
>   * [x]  DefaultError
> * [x]  Ajv is correctly configured with ajv-errors (Edit https://github.com/TypedProject/tsed/blob/production/packages/ajv/src/services/Ajv.ts#L48)
> * [x]  Unit test cover correctly the decorators.
  > * [ ]  Documentation is up-to-date (model.md and ajv.md)
> * [ ]  Integration test cover the success and error cases.

I still have a question about what a typeError represents, is it related to decorators like @Integer() ? 

I will do the documentation once we have finished everything to not have to modify it everytime we change something during the review.

Unit tests could have better coverage.

**Comment by Romakita** [comment]
@Crunchyman-ralph TypeError is a simple shortcut decorator which can be used on class, to customize the error message related to `type` field on json schema ;)

```typescript
@TypeError("should be an object") // shortcut to ErrorMsg({type: 'msg'})
class MyClass {}
```

See you :)

**Comment by Crunchyman-ralph** [comment]
> * [x]   Documentation is up-to-date (model.md and ajv.md)
> * [x]   Integration test cover the success and error cases. 

Thanks for the support on this PR romain, I have added the TypeError decorator and noticed how you could using @ErrorMsg with this format (@ErrorMsg({foo:bar}) instead of @ErrorMsg("foo", bar))

So I just decided to change the way the function works, also did a few more fixes.

All that I have left is the documentation and the integration tests.

For the documentation:
- where would you put this part of the documentation in the model.md file and the ajv.md file (since we noticed that these changes were specific to spec and not necessarily to ajv, I thought maybe it should only go in model.md ?

For the integration tests:
- Where would they go ? Do you have any good examples of any other integration tests by any chance ?

**Comment by Romakita** [comment]
> For the documentation:
> where would you put this part of the documentation in the model.md file and the ajv.md file (since we noticed that these changes were specific to spec and not necessarily to ajv, I thought maybe it should only go in model.md ?
 
We can update the model.md by adding a new section :) (ajv is always included in a Ts.ED app)   

> For the integration tests:
> Where would they go ? Do you have any good examples of any other integration tests by any chance ?
Yes, you have two places for that.

1) If you want to run integration over all platform (Express, Koa), this is the good place:
https://github.com/tsedio/tsed/tree/production/packages/platform/platform-test-utils/src/tests

You have to run `yarn test:integration` - (run it directly from IDE isn't possible).

2) If you code isn't really related to a specific platform, you can add the integration test here:  

https://github.com/tsedio/tsed/tree/production/packages/platform/platform-express/test

I use this location generally to test the integration between `@tsed/common` and `@tsed/schema`/`@tsed/json-mapper` packages :) 

I think the second location will be the good place for this story (ajv are not directly related to express/koa).

See you @Crunchyman-ralph ;)

**Comment by Crunchyman-ralph** [comment]
@Romakita I have finally gotten around to getting the end to end tests as well as the documentation down.

It's a shame the documentation doesn't have a search feature, would make things much easier.

Other than that the PR is ready for review and eventually a merge.

It has been a pleasure contributing to this project, I learned a lot (especially with unit tests and e2e tests)

**Comment by Romakita** [comment]
Hello @Crunchyman-ralph 
Capture d’écran 2023-01-15 à 08 50 38

The documentation have search feature on the website:
 Capture d’écran 2023-01-15 à 08 50 38
 
it use algolia :D 

Thanks again for the PR! awesome!

File: packages/specs/ajv/package.json [hunk 1]
```diff
@@ -28,14 +28,16 @@
     "@tsed/di": "7.10.1",
     "@tsed/exceptions": "7.10.1",
     "@tsed/schema": "7.10.1",
-    "ajv": "8.11.0"
+    "ajv": "8.11.0",
+    "ajv-errors": "3.0.0"
```
Comment (Romakita) [inline packages/specs/ajv/package.json]: you can set ajv-errors as dependencies like ajv-format

**Review by Romakita — CHANGES_REQUESTED** [review]

**Comment by Romakita** [comment]
OK just a minor change around the package.json

We will merge the PR. The next step (next PR) will be to wrap all decorators that produce an error message (like Required decorator).
We can wrap the following decorators:

- [x] MinLength/MaxLength
- [x] Min/Max/ExclusiveMin/ExclusiveMax
- [x] MinItems/MaxItems
- [x] MinProperties/MaxProperties
- [x] UniqItems
- [x] Format
- [x] MultipleOf
- [x] Pattern
- [x] Enum ?

It should be easy, now that the mechanism is implemented :D

**Review by Romakita — APPROVED** [review]

**Pull request merged**